File: app/src/main/java/org/wikipedia/feed/model/Card.java
package org.wikipedia.feed.model;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class Card {
    @NonNull public String title() {
        return "";
    }

    @Nullable public String subtitle() {
        return null;
    }

    @Nullable public Uri image() {
        return null;
    }

    @Nullable public String extract() {
        return null;
    }

    @NonNull public abstract CardType type();

    public void onDismiss() {
    }

    public void onRestore() {
    }

    public String getHideKey() {
        return Long.toString(type().code() + dismissHashCode());
    }

    protected int dismissHashCode() {
        return hashCode();
    }

    @Override
    public int hashCode() {
        return 31 * type().code() + title().hashCode();
    }
}

File: app/src/main/java/org/wikipedia/feed/model/Card.kt
package org.wikipedia.feed.model

abstract class Card {

    val hideKey get() = dismissHashCode().toString()

    open fun moduleKey(): String {
        return ""
    }

    open fun title(): String {
        return ""
    }

    protected open fun dismissHashCode(): Int {
        return hashCode()
    }

    override fun hashCode(): Int {
        return 31 * title().hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Card) return false
        if (hideKey != other.hideKey) return false
        return true
    }
}
